RecoLocalTracker/SiPixelDigiMorphing/plugins/SiPixelDigiMorphing.cc

```diff
@@ -0,0 +1,172 @@
+#include "FWCore/PluginManager/interface/ModuleDef.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "DataFormats/DetId/interface/DetId.h"
+#include "SiPixelDigiMorphing.h"
+#include <bitset>
+
+SiPixelDigiMorphing::SiPixelDigiMorphing(edm::ParameterSet const& conf)
+    : tPutPixelDigi(produces<edm::DetSetVector<PixelDigi>>()),
+      nrows_(conf.getParameter<int32_t>("nrows")),
+      ncols_(conf.getParameter<int32_t>("ncols")),
+      nrocs_(conf.getParameter<int32_t>("nrocs")),
+      iters_(conf.getParameter<int32_t>("iters")),
+      fakeAdc_(conf.getParameter<uint32_t>("fakeAdc")) {
+  tPixelDigi = consumes<edm::DetSetVector<PixelDigi>>(conf.getParameter<edm::InputTag>("src"));
+
+  if (ncols_ % nrocs_ == 0) {
+    ncols_r_ = ncols_ / nrocs_;
+  } else {
+    throw cms::Exception("Configuration") << "[SiPixelDigiMorphing]:"
+                                          << " number of columns not divisible with"
+                                          << " number of ROCs\n";
+  }
+
+  if (ncols_r_ + 2 * iters_ <= int(sizeof(uint64_t) * 8)) {
+    ksize_ = 2 * iters_ + 1;
+  } else {
+    throw cms::Exception("Configuration") << "[SiPixelDigiMorphing]:"
+                                          << " too many columns per ROC"
+                                          << " or too many iterations set\n"
+                                          << " Ncol/Nrocs+2*iters should not be"
+                                          << " more than " << sizeof(uint64_t) << "\n";
+  }
+
+  std::vector<int32_t> k1(conf.getParameter<std::vector<int32_t>>("kernel1"));
+  std::vector<int32_t> k2(conf.getParameter<std::vector<int32_t>>("kernel2"));
+
+  kernel1_.resize(ksize_, 0);
+  kernel2_.resize(ksize_, 0);
+  mask_ = 0;
+  int w = (ncols_r_ + 2 * iters_) / ksize_ + ((ncols_r_ + 2 * iters_) % ksize_ != 0);
+  for (int j = 0; j < w; j++) {
+    for (int ii = 0; ii < ksize_; ii++) {
+      kernel1_[ii] <<= ksize_;
+      kernel1_[ii] |= k1[ii];
+      kernel2_[ii] <<= ksize_;
+      kernel2_[ii] |= k2[ii];
+    }
+    mask_ <<= ksize_;
+    mask_ |= 1;
+  }
+  mask_ <<= iters_;
+}
+
+SiPixelDigiMorphing::~SiPixelDigiMorphing() = default;
+
+void SiPixelDigiMorphing::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("src", edm::InputTag("siPixelDigis"));
+  desc.add<int32_t>("nrows", 160);
+  desc.add<int32_t>("ncols", 416);
+  desc.add<int32_t>("nrocs", 8);
+  desc.add<int32_t>("iters", 1);
+  desc.add<std::vector<int32_t>>("kernel1", {7, 7, 7});
+  desc.add<std::vector<int32_t>>("kernel2", {2, 7, 2});
+  desc.add<uint32_t>("fakeAdc", 100);
+
+  descriptions.add("SiPixelDigiMorphing", desc);
```

Comment:
```suggestion
  descriptions.addDefault(desc);
```
is a preferred

Reply:
Or did you mean`addWithDefaultLabel()`?  (`addDefault()` does not generate `cfi.py` file)

Reply:
> Or did you mean`addWithDefaultLabel()`? (`addDefault()` does not generate `cfi.py` file)

yes; thank you for noticing. I updated the suggestion